### PR TITLE
8261731: shallow copy the internal buffer of a scalar-replaced java.lang.String object

### PR DESCRIPTION
**Pull request by navyxliu**
There are 3 nodes involving in the construction of a java.lang.String object.
1. Allocate of itself, aka. alloc
2. AllocateArray of a byte array, which is value:byte[], aka. aa
3. ArrayCopyNode which copys in the contents of value, aka. ac

Lemma
When a String object `alloc` is scalar replaced, C2 can eliminate `aa` and `ac`.

Because `alloc` is scalar replaced, it must be non-escaped. The field value:byte[] of j.l.String cannot be seen by external world, therefore it must not be global escaped. Because the buffer is marked as stable, it is safe to assume its contents are whatever ac copies in. Because all public java.lang.String constructors clone the incoming array, the source of `ac` is stable as well.

It is possible to rewire `aa` to the source of ac with the correct offset. That is to say, we can replace both `aa` and `ac` with a “shallow copy” of the source of `ac`. It’s safe if C2 keeps a reference of the source oop for all safepoints.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Issue
 * [JDK-8261731](https://bugs.openjdk.java.net/browse/JDK-8261731): shallow copy the internal buffer of a scalar-replaced java.lang.String object

### Download
To checkout this PR locally:
`$ git fetch https://git.openjdk.java.net/jdk pull/2570/head:pull/2570`
`$ git checkout pull/2570`

To update a local copy of the PR:
`$ git checkout pull/2570`
`$ git pull https://git.openjdk.java.net/jdk pull/2570/head`

### CODE REVIEW

**Comment by reinrich** [comment]
Hi,

this is a smart optimization.

> 
> 
> There are 3 nodes involving in the construction of a java.lang.String object.
> 
>     1. Allocate of itself, aka. alloc
> 
>     2. AllocateArray of a byte array, which is value:byte[], aka. aa
> 
>     3. ArrayCopyNode which copys in the contents of value, aka. ac
> 
> 
> Lemma
> When a String object `alloc` is scalar replaced, C2 can eliminate `aa` and `ac`.
> 
> Because `alloc` is scalar replaced, it must be non-escaped. The field value:byte[] of j.l.String cannot be seen by external world, therefore it must not be global escaped. Because the buffer is marked as stable, it is safe to assume its contents are whatever ac copies in. Because all public java.lang.String constructors clone the incoming array, the source of `ac` is stable as well.

Are you saying that the source of `ac` cannot be accessed by another thread because of the cloning in the constructor? But the resulting string instance which is used to construct the non-escape instance can be GlobalEscape and then the source of `ac` is accessible to other threads, isn't it?

**Comment by navyxliu** [comment]
> Hi,
> 
> this is a smart optimization.
> 
> > There are 3 nodes involving in the construction of a java.lang.String object.
> > ```
> > 1. Allocate of itself, aka. alloc
> > 
> > 2. AllocateArray of a byte array, which is value:byte[], aka. aa
> > 
> > 3. ArrayCopyNode which copys in the contents of value, aka. ac
> > ```
> > 
> > 
> > Lemma
> > When a String object `alloc` is scalar replaced, C2 can eliminate `aa` and `ac`.
> > Because `alloc` is scalar replaced, it must be non-escaped. The field value:byte[] of j.l.String cannot be seen by external world, therefore it must not be global escaped. Because the buffer is marked as stable, it is safe to assume its contents are whatever ac copies in. Because all public java.lang.String constructors clone the incoming array, the source of `ac` is stable as well.
> 
> Are you saying that the source of `ac` cannot be accessed by another thread because of the cloning in the constructor? But the resulting string instance which is used to construct the non-escape instance can be GlobalEscape and then the source of `ac` is accessible to other threads, isn't it?

Hi, @reinrich 
You might not know, but I learn how to reallocate an object in deoptimization from your previous patches. Thank you!

You are right. The source of ac (`src`) might be escaped.  I didn't say other threads can't access it. I said we need to guarantee the `src` is stable,  or  it's a "frozen" array in JDK-8261007. 

To be honest, I didn't check the src is frozen. In practice, I only see ArrayCopy in construction [here](https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/StringLatin1.java#L769).This method is creating a new substring from an established array, so its value is `stable`. 

After I read your comment, I went through String.java. I do find one open-ended constructor. yes, it's a problem!
https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/String.java#L395

If the frozen attribute is not present, what I can come up with. an array is "stable". What's do you think?
1) has annotation "stable" or
2) it's non-escaped and
3) can't find any store nodes along its mem stream.

**Comment by reinrich** [comment]
Hi @navyxliu,

> You might not know, but I learn how to reallocate an object in deoptimization from your previous patches. Thank you!

Oh, great! :)

> You are right. The source of ac (`src`) might be escaped. I didn't say other threads can't access it. I said we need to guarantee the `src` is stable, or it's a "frozen" array in JDK-8261007.
> 
> To be honest, I didn't check the src is frozen. In practice, I only see ArrayCopy in construction [here](https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/StringLatin1.java#L769).This method is creating a new substring from an established array, so its value is `stable`.

Sorry, I wasn't aware of the @Stable annotation on the value array.

https://github.com/openjdk/jdk/blob/d19503353e5c347ce393544a3a30d5caec53d133/src/java.base/share/classes/java/lang/String.java#L154

So my concern was that another thread could for example use reflection to modify
the value array but I reckon this is illegal then (I wonder if it is checked in
reflection...).

Also there is already UseStringDeduplication that relies on the value array being stable.

> After I read your comment, I went through String.java. I do find one open-ended constructor. yes, it's a problem!
> https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/String.java#L395

I see this one is deprecated. Why do you think there's a problem?

> If the frozen attribute is not present, what I can come up with. an array is "stable". What's do you think?
> 
>     1. has annotation "stable" or
> 
>     2. it's non-escaped and
> 
>     3. can't find any store nodes along its mem stream.

Currently I'd think this is not necessary. `@Stable` is strong enough.

The optimization reminds me a bit of earlier Java versions where String
instances had an offset field and the value array could be shared among String
instances.

I guess the new String needs to be scalarized mostly because you cannot take
care of the offset otherwise.

Coincidentally I'm currently looking at
[LoadNode::can_see_arraycopy_value()](https://github.com/openjdk/jdk/blob/b955f85e03bafe8ce39677d0af06bf1ceb7e2cbb/src/hotspot/share/opto/memnode.cpp#L951)
which does something similar you want to do. There I don't see any concerns
about the src being stable. In fact I don't think this is correct. Just a side
mark...

These are my (somewhat random) thoughts so far. I'd think it is a legal and useful optimization (though haven't yet looked at the code yet).

Thanks, Richard.

File: src/hotspot/share/opto/macro.cpp [hunk 1]
```diff
@@ -1054,6 +1217,206 @@ void PhaseMacroExpand::process_users_of_allocation(CallNode *alloc) {
...
+//                -----------
+//                |  AddP   |
+//                -----------
+//                /          \
+//               ----------   ----------
+//               | LoadUB  |  |LoadRange|
+//               ----------   ----------
+//
+//
+//  EncodeP:  delele because we don't need storeN
```
Comment (simonis) [inline src/hotspot/share/opto/macro.cpp:1317]: "delete" not "delele"

**Comment by simonis** [comment]
What about GC? What happens if the original String isn't reachable any more? Do you put a reference to the byte array into the corresponding oop maps to make sure it can't be garbage collected?

File: src/hotspot/share/opto/macro.cpp (hunk 1 above)
Comment (navyxliu) [reply src/hotspot/share/opto/macro.cpp:1317]: got it. thanks. I will update it.

**Comment by navyxliu** [comment]
> What about GC? What happens if the original String isn't reachable any more? Do you put a reference to the byte array into the corresponding oop maps to make sure it can't be garbage collected?

hi, Volker, 

I don't know OopMap very much.  IMHO, it should be discovered by GC because I put the byte array oop to `jvm->locals`
I do that for all safepoints nodes. 

```
          // 1. src oop
          sfpt->add_req(ac->in(ArrayCopyNode::Src));
```
https://github.com/openjdk/jdk/pull/2570/files#diff-2faebd05d08f9115f8d9ef771644cf05087a6986c2f9013d7163c6aa720169c3R995

**Comment by simonis** [comment]
Hi Xin,
when building a fastdebug build of your latest changes I got a crash because of an assertion in your new code.
I've attached the hs_err and replay file. After restarting the build it succeeded so there must have been a special situation. Hopefully the replay file can help you to reproduce the error.

Best regards,
Volker

[hs_err_pid18140.log](https://github.com/openjdk/jdk/files/6075724/hs_err_pid18140.log)
[replay_pid18140.log](https://github.com/openjdk/jdk/files/6075725/replay_pid18140.log)

**Comment by navyxliu** [comment]
> Hi Xin,
> when building a fastdebug build of your latest changes I got a crash because of an assertion in your new code.
> I've attached the hs_err and replay file. After restarting the build it succeeded so there must have been a special situation. Hopefully the replay file can help you to reproduce the error.
> 
> Best regards,
> Volker
> 
> [hs_err_pid18140.log](https://github.com/openjdk/jdk/files/6075724/hs_err_pid18140.log)
> [replay_pid18140.log](https://github.com/openjdk/jdk/files/6075725/replay_pid18140.log)

hi, Volker, 
thank you for using this patch. 
Yes, indeed, it's a bug. I introduced yesterday, I got a silly typo for the AddP nodes.  I need to use else if instead of else. 
in your replay file, optimization met a special node StrEquals. I have handled it above. 
```
diff --git a/src/hotspot/share/opto/macro.cpp b/src/hotspot/share/opto/macro.cpp
index 56dab5c6449..76e97087fd9 100644
--- a/src/hotspot/share/opto/macro.cpp
+++ b/src/hotspot/share/opto/macro.cpp
@@ -1372,7 +1372,7 @@ void PhaseMacroExpand::process_users_of_string_allocation(AllocateArrayNode* all
             if (n->in(3) == use) { // str2
               _igvn.replace_input_of(n, 3, dst_adr);
             }
-          } if (n->is_AddP()) {
+          } else if (n->is_AddP()) {
             // Skip second AddP. This node must be handled by the upper level.
           } else {
             assert(n->Opcode() == Op_LoadUB || n->Opcode() == Op_LoadB, "unknow code shape");
```
I will fix soon.

**Comment by TobiHartmann** [comment]
I didn't look at the code yet but run this through some quick testing. The following tests fail with `-ea -esa -XX:CompileThreshold=100:`
- compiler/codecache/stress/RandomAllocationTest.java
- compiler/codecache/stress/UnexpectedDeoptimizationTest.java
```
#  Internal Error (open/src/hotspot/share/opto/node.hpp:825), pid=27269, tid=27288
#  assert(is_AllocateArray()) failed: invalid node class: IfTrue

Current CompileTask:
C2:  12403 2987       4       java.lang.Class::descriptorString (160 bytes)

Stack: [0x00007fc080bfc000,0x00007fc080cfd000],  sp=0x00007fc080cf8340,  free space=1008k
Native frames: (J=compiled Java code, A=aot compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x1302c8c]  PhaseMacroExpand::scalar_replacement(AllocateNode*, GrowableArray<SafePointNode*>&)+0x88c
V  [libjvm.so+0x1303e5e]  PhaseMacroExpand::eliminate_allocate_node(AllocateNode*) [clone .part.0]+0x50e
V  [libjvm.so+0x130447c]  PhaseMacroExpand::eliminate_macro_nodes()+0x57c
V  [libjvm.so+0xa0fbb5]  Compile::Optimize()+0x1235
V  [libjvm.so+0xa11c85]  Compile::Compile(ciEnv*, ciMethod*, int, bool, bool, bool, bool, DirectiveSet*)+0x1905
V  [libjvm.so+0x83d69a]  C2Compiler::compile_method(ciEnv*, ciMethod*, int, bool, DirectiveSet*)+0x1ea
V  [libjvm.so+0xa21aa1]  CompileBroker::invoke_compiler_on_method(CompileTask*)+0xf21
V  [libjvm.so+0xa22748]  CompileBroker::compiler_thread_loop()+0x5a8
V  [libjvm.so+0x18488e1]  JavaThread::thread_main_inner()+0x271
V  [libjvm.so+0x1850b30]  Thread::call_run()+0x100
V  [libjvm.so+0x153b086]  thread_native_entry(Thread*)+0x116
```

**Comment by TobiHartmann** [comment]
java/lang/String/Split.java and java/lang/management/MemoryMXBean/LowMemoryTest2.sh fail with:
```
#  Internal Error (open/src/hotspot/share/opto/macro.cpp:1378), pid=5320, tid=5499
#  assert(n->Opcode() == Op_LoadUB || n->Opcode() == Op_LoadB) failed: unknow code shape

Current CompileTask:
C2:  11623  994   !   4       Split::main (651 bytes)

Stack: [0x00007ff1869fe000,0x00007ff186aff000],  sp=0x00007ff186afa2d0,  free space=1008k
Native frames: (J=compiled Java code, A=aot compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x12f8a5f]  PhaseMacroExpand::process_users_of_string_allocation(AllocateArrayNode*, ArrayCopyNode*)+0x83f
V  [libjvm.so+0x1303e0e]  PhaseMacroExpand::eliminate_allocate_node(AllocateNode*) [clone .part.0]+0x4be
V  [libjvm.so+0x130447c]  PhaseMacroExpand::eliminate_macro_nodes()+0x57c
V  [libjvm.so+0xa0fbb5]  Compile::Optimize()+0x1235
V  [libjvm.so+0xa11c85]  Compile::Compile(ciEnv*, ciMethod*, int, bool, bool, bool, bool, DirectiveSet*)+0x1905
V  [libjvm.so+0x83d69a]  C2Compiler::compile_method(ciEnv*, ciMethod*, int, bool, DirectiveSet*)+0x1ea
V  [libjvm.so+0xa21aa1]  CompileBroker::invoke_compiler_on_method(CompileTask*)+0xf21
V  [libjvm.so+0xa22748]  CompileBroker::compiler_thread_loop()+0x5a8
V  [libjvm.so+0x18488e1]  JavaThread::thread_main_inner()+0x271
V  [libjvm.so+0x1850b30]  Thread::call_run()+0x100
V  [libjvm.so+0x153b086]  thread_native_entry(Thread*)+0x116
``` 

And I see some internal testing failing with SIGSEGV in `PhaseMacroExpand::eliminate_allocate_node`

**Comment by TobiHartmann** [comment]
Yes, I've used the latest version for testing.

**Comment by navyxliu** [comment]
> Split.java

hi, @TobiHartmann , 
Could you teach me how to trigger crashes for java/lang/String/Split.java and java/lang/management/MemoryMXBean/LowMemoryTest2.sh ? 

I tried different options like "'-ea -esa -XX:CompileThreshold=100' with/without -XComp, -XX:-TieredCompilation. still no luck.
or Could you paste a replay file so I can investigate it?

**Comment by TobiHartmann** [comment]
Sure, here are the replay and hs_err files for the Split.java failure:
[replay_pid5320.log](https://github.com/openjdk/jdk/files/6106543/replay_pid5320.log)
[hs_err_pid5320.log](https://github.com/openjdk/jdk/files/6106544/hs_err_pid5320.log)

And for the LowMemoryTest2 failure:
[replay_pid18829.log](https://github.com/openjdk/jdk/files/6106550/replay_pid18829.log)
[hs_err_pid18829.log](https://github.com/openjdk/jdk/files/6106551/hs_err_pid18829.log)

It seems like the failures are very intermittent though :(

**Comment by navyxliu** [comment]
hi, @TobiHartmann , 
Could you try out the latest version 94626e8? I haven't  reproduced the error from the two replay files.  My best guess is that both of them crashed because of some special nodes which are subclasses of StrIntrinsicNode. I only handled the case StrEqualsNode before. This patch handles them.  It also dumps the unknown node to tty in case it still meets unknown code shape.

**Comment by TobiHartmann** [comment]
Sure, I'll re-run testing.

After looking through the changes, I'm concerned about all the complexity that is added to handle this special case. Do you have any performance numbers other than microbenchmark results that would justify this?

**Comment by TobiHartmann** [comment]
The changes to macro.cpp do not build anymore after [JDK-8263577](https://bugs.openjdk.java.net/browse/JDK-8263577) because the signature of `Node::replace_edges_in_range` changed.

**Comment by navyxliu** [comment]
> Sure, I'll re-run testing.
> 
> After looking through the changes, I'm concerned about all the complexity that is added to handle this special case. Do you have any performance numbers other than microbenchmark results that would justify this?

hi, @TobiHartmann , 
It's not my intention to make a big patch.  I would like to get help from reviewers. Am I making thing over complex? Is there an alternative to make thing simpler? 

The reason that the patch is big because I meet some constraints in current optimizing compiler. 
1. It seems all oops must be allocated on heap. There's no pseudo oop. or I can’t find a way to create a temp of fake oop.  
2. Scalar_replacement and ArrayCopy idealization don't support variable-length array(VLA).  String::value is a VLA. 
3. Lack supports of frozen array in C2's typing system. JDK-8261007
4. re-allocation doesn't support nested objects. j.l.String has 4 fields. String::value is an array of byte, which I need to reallocate too.

To overcome the constraints, I make the following new features in this patch.
1. `process_users_of_string_allocation` intercepts all uses of AllocateArray. 
2. reorder the trinity Allocate-AllocateArray-ArrayCopy(or alloc-aa-ac) and eliminate them in MacroExpansion phase.
3. `ArrayCopyNode::is_string_copy` conservatively recognize string initialization from another string. 
4. invent a scheme to make deoptimization support a nested object, so deopt can reallocation a j.l.String object.

TBH, I haven't seen obvious performance improvement in SpebJBB. Only microbench can prove my idea.  When I profile a string intensive application, I found that the major blocker is JDK-8261531.  Current implementation requires the String object must be scalar-replacement first.  A String object is NOT scalar replaceable until its length is a compiler-time constant.  It's pretty rare in real world. Therefore, I would like to put the evaluation of performance impact on hold before I resolve JDK-8261531.

Could you help me verify that the current code can handle all code shape first?  Further, I really appreciate that you can give me some advice in terms of high-level design.

thanks,
--lx

**Comment by TobiHartmann** [comment]
Hi @navyxliu,

yes, these are all well-known limitations and unfortunately there is no simple solution to get rid of them. But instead of adding code to work around the limitations in special cases, I think it would be better to find more general solutions. Especially, if there are no performance numbers that would justify the amount of code complexity for handling that special case.

And as you've said, even with your patch, [JDK-8261531](https://bugs.openjdk.java.net/browse/JDK-8261531) is still a major blocker and I think that alone is very complex to fix.

But that's only my opinion, let's see what other reviewers think.

Best regards,
Tobias

**Comment by TobiHartmann** [comment]
Hi @navyxliu,

yes, these are all well-known limitations and unfortunately there is no simple solution to get rid of them. But instead of adding code to work around the limitations in special cases, I think it would be better to find more general solutions. Especially, if there are no performance numbers that would justify the amount of code complexity for handling that special case.

And as you've said, even with your patch, [JDK-8261531](https://bugs.openjdk.java.net/browse/JDK-8261531) is still a major blocker and I think that alone is very complex to fix.

But that's only my opinion, let's see what other reviewers think.

Best regards,
Tobias

**Comment by vnkozlov** [comment]
@navyxliu I think you want to fix a lot of issues once. You need to split problems you see.

I did few experiments and the one issue which could be fixed first is scalarizing of array created by `Arrays.copyOfRange()` used by `String::substring()`.

I see that current JDK easy scalarize new array and arraycopy():
```
     private static int test4(byte[] src) {
         byte[] dst = new byte[6];
         System.arraycopy(src, 1, dst, 0, 6);
         return dst[2] + dst[5];
     }
```
But it did not scalarize:
```
     private static int test5(byte[] src) {
         byte[] dst = Arrays.copyOfRange(src, 1, 7);
         return dst[2] + dst[5];
     }
```
Even so code in `Arrays.copyOfRange()` is the same as in `test4()`:
https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/util/Arrays.java#L3817

Arrays does not escaped but it is not eliminated because Loads are not redirected into `dst` array as in `test4()` case.

I think this should be fixed first.

**Comment by vnkozlov** [comment]
Hmm, actually it was easy solved with `-XX:DisableIntrinsic=_min` :(
We should not implement intrinsics for everything!!!!!

**Comment by cl4es** [comment]
You seem to have had a run-in with these intrinsics causing issues before: https://bugs.openjdk.java.net/browse/JDK-8039104 - is this a similar issue?

On a trivial micro adding together a few Math.min values the intrinsic seem to have no effect on my x64 system, down to generating the exact same assembly.

**Comment by vnkozlov** [comment]
Yes, my test5() have `cmove`:
```
021     cmovlge RBP, RDX        # signed, int
```

**Comment by vnkozlov** [comment]
Wow, thanks @cl4es. I forgot about that bug.

**Pull request closed without merging**